ambry-tools/src/main/java/com.github.ambry/store/StoreCopier.java

```diff
@@ -0,0 +1,278 @@
+   * @param metricRegistry {@link MetricRegistry} to use for metrics.
+   * @param storeKeyFactory the {@link StoreKeyFactory} to use for {@link StoreKey}s in the {@link Store}.
+   * @param diskIOScheduler the {@link DiskIOScheduler} to use.
+   * @param transformers the list of {@link Transformer} functions to execute. They will be executed in order.
+   * @param time the {@link Time} instance to use.
+   * @throws StoreException
+   */
+  public StoreCopier(String storeId, File srcDir, File tgtDir, long storeCapacity, long fetchSizeInBytes,
+      StoreConfig storeConfig, MetricRegistry metricRegistry, StoreKeyFactory storeKeyFactory,
+      DiskIOScheduler diskIOScheduler, List<Transformer> transformers, Time time) throws StoreException {
+    this.storeId = storeId;
+    this.fetchSizeInBytes = fetchSizeInBytes;
+    this.transformers = transformers;
+    StorageManagerMetrics metrics = new StorageManagerMetrics(metricRegistry);
+    MessageStoreRecovery recovery = new BlobStoreRecovery();
+    src = new BlobStore(storeId, storeConfig, null, null, diskIOScheduler, metrics, srcDir.getAbsolutePath(),
+        storeCapacity, storeKeyFactory, recovery, null, time);
+    tgt = new BlobStore(storeId + "_tmp", storeConfig, scheduler, null, diskIOScheduler, metrics,
+        tgtDir.getAbsolutePath(), storeCapacity, storeKeyFactory, recovery, null, time);
+    src.start();
+    tgt.start();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (!isOpen) {
+      return;
+    }
+    try {
+      scheduler.shutdown();
+      if (!scheduler.awaitTermination(1, TimeUnit.SECONDS)) {
+        logger.error("Could not shut down scheduler");
+      }
+      src.shutdown();
+      tgt.shutdown();
+      isOpen = false;
+    } catch (Exception e) {
+      throw new IOException(e);
+    }
+  }
+
+  /**
+   * Copies data starting from {@code startToken} until all the data is copied.
+   * @param startToken the {@link FindToken} to start copying from. Does not perform any duplication checks at
+   *                   destination.
+   * @return the {@link FindToken} until which data has been copied.
+   * @throws IOException if there is any I/O error while copying.
+   * @throws StoreException if there is any exception dealing with the stores.
+   */
+  public FindToken copy(FindToken startToken) throws IOException, StoreException {
+    FindToken lastToken = null;
+    FindToken token = startToken;
+    do {
+      lastToken = token;
+      FindInfo findInfo = src.findEntriesSince(lastToken, fetchSizeInBytes);
+      List<MessageInfo> messageInfos = findInfo.getMessageEntries();
+      for (MessageInfo messageInfo : messageInfos) {
+        logger.trace("Processing {} - isDeleted: {}, isExpired {}", messageInfo.getStoreKey(), messageInfo.isDeleted(),
+            messageInfo.isExpired());
+        if (!messageInfo.isExpired() && !messageInfo.isDeleted()) {
+          if (messageInfo.getSize() > Integer.MAX_VALUE) {
```

Comment:
why do we have this check? Surely blobs within the store cannot be of `size > Integer.MAX_VALUE`?

Reply:
Abundance of caution I guess? I don't have a slam dunk reason for it - Just wanted to make sure that the cast in the next line is a safe cast.